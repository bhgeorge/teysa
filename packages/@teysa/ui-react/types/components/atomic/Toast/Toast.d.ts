declare type ToastVariant = 'success' | 'error' | 'none';
export interface ToastProps {
    text?: string;
    title?: string;
    variant?: ToastVariant;
    role?: 'alert';
}
declare const Toast: ({ title, text, variant, role }: ToastProps) => JSX.Element;
export default Toast;
