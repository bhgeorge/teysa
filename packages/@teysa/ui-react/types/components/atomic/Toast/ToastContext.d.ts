import React from 'react';
import { ToastProps } from './Toast';
interface ToastContextProviderProps {
    children: React.ReactNode | React.ReactNode[];
}
export declare const useToastContext: () => {
    addToast: (toast: ToastProps) => void;
    test: string;
};
declare const ToastContextProvider: ({ children }: ToastContextProviderProps) => JSX.Element;
export default ToastContextProvider;
