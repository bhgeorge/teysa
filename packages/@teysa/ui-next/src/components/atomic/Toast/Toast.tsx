// Global
import React from 'react';
import classNames from 'classnames';
// Components
import { Heading } from '../Heading/Heading';

type ToastVariant = 'success' | 'warning' | 'error' | 'none';

export interface ToastProps {
  text?: string;
  title?: string;
  variant?: ToastVariant;
  role?: 'alert';
}

const variantClasses: Record<ToastVariant, string> = {
  success: 'border-l-theme-success',
  warning: 'border-l-theme-warning',
  error: 'border-l-theme-error',
  none: '',
};

export function Toast({ title, text, variant = 'none', role }: ToastProps) {
  if (!title && !text) {
    return <></>;
  }

  return (
    <div
      className={classNames(
        'bg-theme-bg',
        'border',
        'border-l-8',
        'border-theme-bg-alt',
        'p-4',
        'shadow-xl',
        'w-full',
        variantClasses[variant]
      )}
      role={role}
    >
      {!!title && (
        <Heading
          headingLevel={2}
          size="xs"
          className={!!text ? 'mb-1' : undefined}
        >
          {title}
        </Heading>
      )}
      {!!text && <p className="text-xs">{text}</p>}
    </div>
  );
}

export default Toast;
