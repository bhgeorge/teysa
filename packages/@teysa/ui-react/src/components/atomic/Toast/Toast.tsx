// Global
import React from 'react';
import classNames from 'classnames';
// Components
import Heading from '../Heading/Heading';

type ToastVariant = 'success' | 'error' | 'none';

interface ToastProps {
  text?: string;
  title?: string;
  variant?: ToastVariant;
}

const variantClasses: Record<ToastVariant, string> = {
  success: 'border-l-theme-success',
  error: 'border-l-theme-error',
  none: '',
};

const Toast = ({ title, text, variant = 'none' }: ToastProps) => {
  if (!title && !text) {
    return <></>;
  }

  return (
    <div
      className={classNames(
        'bg-theme-bg',
        'border',
        'border-l-8',
        'border-theme-bg-alt',
        'p-4',
        'shadow-xl',
        'w-full',
        variantClasses[variant]
      )}
    >
      {!!title && (
        <Heading headingLevel={2} size="xs" className={!!text ? 'mb-1' : undefined}>
          {title}
        </Heading>
      )}
      {!!text && <p className="text-xs">{text}</p>}
    </div>
  );
};

export default Toast;
