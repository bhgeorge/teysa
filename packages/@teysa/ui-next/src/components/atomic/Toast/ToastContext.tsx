// Global
import React, { useState, createContext, useContext, useId } from 'react';
import { TransitionGroup, CSSTransition } from 'react-transition-group';
// Components
import { Container } from '../../layout/Container/Container';
// Local
import { Toast, ToastProps } from './Toast';

interface GlobalToastProps extends ToastProps {
  id: string;
}

interface ToastContextProviderProps {
  children: React.ReactNode | React.ReactNode[];
}

const ToastContext = createContext({
  addToast: (toast: ToastProps) => {
    console.log(toast);
  },
  test: 'name',
});

export const useToastContext = () => useContext(ToastContext);

const TIMEOUT = 2500;

export function ToastContextProvider({ children }: ToastContextProviderProps) {
  const [numToasts, setNumToasts] = useState<number>(0);
  const [toasts, setToasts] = useState<GlobalToastProps[]>([]);

  const componentId = useId();

  const removeToast = (id: string) => {
    setToasts(toasts => toasts.filter(toast => toast.id !== id));
  };

  const addToast = (toast: ToastProps) => {
    const id = `${componentId}__${numToasts}`;
    setNumToasts(numToasts + 1);

    window.setTimeout(() => {
      removeToast(id);
    }, TIMEOUT);

    setToasts(toasts => [...toasts, { ...toast, id }]);
  };

  return (
    <ToastContext.Provider value={{ addToast, test: 'moo' }}>
      <div className="fixed top-0 inset-x-0 z-40 p-2 pointer-events-none">
        <Container>
          <ul>
            <TransitionGroup>
              {toasts.map(toast => (
                <CSSTransition
                  key={toast.id}
                  timeout={300}
                  className="animate-drop"
                >
                  <li className="w-full">
                    <Toast
                      title={toast.title}
                      text={toast.text}
                      variant={toast.variant}
                      role="alert"
                    />
                  </li>
                </CSSTransition>
              ))}
            </TransitionGroup>
          </ul>
        </Container>
      </div>
      {children}
    </ToastContext.Provider>
  );
}
