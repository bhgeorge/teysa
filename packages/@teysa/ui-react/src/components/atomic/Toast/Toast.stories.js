// Local
import Toast from './Toast';

export default {
  title: 'Atomic/Toast',
  component: Toast,
};

const Template = (props) => <Toast {...props} />;

export const Default = Template.bind({});
Default.args = {
  title: 'On death trigger',
  text: 'Trigger each of these an additional time',
};
