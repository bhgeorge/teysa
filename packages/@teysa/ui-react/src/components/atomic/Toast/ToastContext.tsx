// Global
import React, { useState, createContext, useContext } from 'react';
import { v4 as uuidv4 } from 'uuid';
import { TransitionGroup, CSSTransition } from 'react-transition-group';
// Components
import Container from '../../layout/Container/Container';
// Local
import Toast, { ToastProps } from './Toast';

interface GlobalToastProps extends ToastProps {
  id: string;
}

interface ToastContextProviderProps {
  children: React.ReactNode | React.ReactNode[];
}

const ToastContext = createContext({
  addToast: (toast: ToastProps) => {},
  test: 'name',
});

export const useToastContext = () => useContext(ToastContext);

const TIMEOUT = 2500;

const ToastContextProvider = ({ children }: ToastContextProviderProps) => {
  const [toasts, setToasts] = useState<GlobalToastProps[]>([]);

  const removeToast = (id: string) => {
    setToasts((toasts) => toasts.filter((toast) => toast.id !== id));
  };

  const addToast = (toast: ToastProps) => {
    const id = uuidv4();

    window.setTimeout(() => {
      removeToast(id);
    }, TIMEOUT);

    setToasts((toasts) => [...toasts, { ...toast, id }]);
  };

  return (
    <ToastContext.Provider value={{ addToast, test: 'moo' }}>
      <div className="fixed top-0 inset-x-0 z-40 p-2 pointer-events-none">
        <Container>
          <ul>
            <TransitionGroup>
              {toasts.map((toast) => (
                <CSSTransition key={toast.id} timeout={300} className="animate-drop">
                  <li className="w-full">
                    <Toast
                      title={toast.title}
                      text={toast.text}
                      variant={toast.variant}
                      role="alert"
                    />
                  </li>
                </CSSTransition>
              ))}
            </TransitionGroup>
          </ul>
        </Container>
      </div>
      {children}
    </ToastContext.Provider>
  );
};

export default ToastContextProvider;
