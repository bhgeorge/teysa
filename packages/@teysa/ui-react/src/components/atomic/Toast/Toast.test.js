// Global
import React from 'react';
import { render } from '@testing-library/react';
// Local
import Toast from './Toast';

const TITLE = 'Teysa';
const TEXT = 'Double on death triggers';

const snapshot = (props) => {
  const component = render(<Toast {...props} />);
  expect(component.container.firstChild).toMatchSnapshot();
};

it('renders with a title', () => {
  snapshot({ title: TITLE });
});

it('renders with text', () => {
  snapshot({ text: TEXT });
});

it('renders with both', () => {
  snapshot({ title: TITLE, text: TEXT });
});

it('does not render with neither', () => {
  snapshot({});
});

it('renders success variant', () => {
  snapshot({ title: TITLE, variant: 'success' });
});

it('renders error variant', () => {
  snapshot({ title: TITLE, variant: 'error' });
});
