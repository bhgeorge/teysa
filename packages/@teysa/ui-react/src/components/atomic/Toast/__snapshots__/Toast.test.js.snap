// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not render with neither 1`] = `null`;

exports[`renders error variant 1`] = `
<div
  class="bg-theme-bg border border-l-8 border-theme-bg-alt p-4 shadow-xl w-full border-l-theme-error"
>
  <h2
    class="text-sm font-bold mb-2"
  >
    Teysa
  </h2>
</div>
`;

exports[`renders success variant 1`] = `
<div
  class="bg-theme-bg border border-l-8 border-theme-bg-alt p-4 shadow-xl w-full border-l-theme-success"
>
  <h2
    class="text-sm font-bold mb-2"
  >
    Teysa
  </h2>
</div>
`;

exports[`renders with a title 1`] = `
<div
  class="bg-theme-bg border border-l-8 border-theme-bg-alt p-4 shadow-xl w-full"
>
  <h2
    class="text-sm font-bold mb-2"
  >
    Teysa
  </h2>
</div>
`;

exports[`renders with both 1`] = `
<div
  class="bg-theme-bg border border-l-8 border-theme-bg-alt p-4 shadow-xl w-full"
>
  <h2
    class="text-sm font-bold mb-2"
  >
    Teysa
  </h2>
  <p
    class="text-xs"
  >
    Double on death triggers
  </p>
</div>
`;

exports[`renders with text 1`] = `
<div
  class="bg-theme-bg border border-l-8 border-theme-bg-alt p-4 shadow-xl w-full"
>
  <p
    class="text-xs"
  >
    Double on death triggers
  </p>
</div>
`;
